### PR TITLE
gh-175 Doc update

### PR DESCRIPTION
**Pull request by GCHQDev404**
gh-175 Example

gh-175 CurrentTimeExample

gh-175 GunzipExample

gh-175 IncrementExample

gh-175 ToBytesExample

gh-175 ToTupleExample

gh-175 MapToTupleExample

gh-175 ToDateStringExample

gh-175 CsvToMapsExample

gh-175 ParseDateExample

gh-175 ParseTimeExample

gh-175 CsvLinesToMapsExample

gh-175 DeserialiseJsonExample

gh-175 ApplyBiFunction

gh-175 FunctionChainExample

gh-175 DeserialiseXmlExample

gh-175 ToTupleExample

### CODE REVIEW

File: src/main/java/uk/gov/gchq/gaffer/doc/function/CurrentTimeExample.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    protected void runExamples() {
+        getCurrentTime();
+    }
+
+    private void getCurrentTime() {
+        // ---------------------------------------------------------
+        final CurrentTime currentTime = new CurrentTime();
+        // ---------------------------------------------------------
+
+        runExample(currentTime, null);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/CurrentTimeExample.java]: This doesn't look right. You still need to give it some test inputs otherwise an empty table gets rendered on the screen:

Input Type | Input | Result Type | Result
-- | -- | -- | --

File: src/main/java/uk/gov/gchq/gaffer/doc/function/IncrementExample.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        addToFloat();
+        addToLong();
+    }
+
+    private void addToInt() {
+        // ---------------------------------------------------------
+        final Increment increment = new Increment(3);
+        // ---------------------------------------------------------
+
+        runExample(increment, "return will match input type int", 2);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/IncrementExample.java]: This creates a section for each one, rather than a row in the table.
Try this instead:
```suggestion
        runExample(increment, null, 2, 2.0d, 2.0f, 2L)
```
Then get rid of the ones below

File: src/main/java/uk/gov/gchq/gaffer/doc/function/IncrementExample.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        runExample(increment, "return will match input type double", 2.0d);
+        runExample(increment, "return will match input type float", 2.0f);
+        runExample(increment, "return will match input type long", 2L);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/IncrementExample.java]: Might be a good idea to add unexpected values like Strings or nulls.

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ToBytesExample.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    protected void runExamples() {
+        stringToBytes();
+    }
+
+    private void stringToBytes() {
+        // ---------------------------------------------------------
+        final ToBytes toBytes = new ToBytes(StandardCharsets.UTF_16);
+        // ---------------------------------------------------------
+
+        runExample(toBytes, null, "example String");
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ToBytesExample.java]: This prints the following:

Input Type | Input | Result Type | Result
-- | -- | -- | --
java.lang.String | example String | [B | [B@69e4de5c

I think you may need a way of displaying byte arrays better

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ToTupleExample.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        // ---------------------------------------------------------
+
+        runExample(function, null, Lists.newArrayList(1, 2, 3, 4));
+        runExample(function, null, new int[]{1, 2, 3, 4});
+        runExample(function, null, new Integer[]{1, 2, 3, 4});
+        Map<String, Object> input = new HashMap<>();
+        input.put("A", 1);
+        input.put("B", 2);
+        input.put("C", 3);
+        runExample(function, null, input);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ToTupleExample.java]: Instead of having three runExample method calls, use a single call with multiple inputs (See the IncrementExample suggestion). This will mean it gets rendered in one table rather than spread out.

File: src/main/java/uk/gov/gchq/gaffer/doc/function/MapToTupleExample.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    private void mapToTuple() {
+        // ---------------------------------------------------------
+        final MapToTuple<String> function = new MapToTuple<>();
+        // ---------------------------------------------------------
+        Map<String, Object> input = new HashMap<>();
+        input.put("A", 1);
+        input.put("B", 2);
+        input.put("C", 3);
+
+        runExample(function, null, input);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/MapToTupleExample.java]: Add more inputs here: try different map types, Maps of generics, FrequencyMaps, nulls etc

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ToDateStringExample.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        toDateStringShortFormattedToMin();
+        toDateStringShortFormattedToDay();
+    }
+
+    private void toDateStringMicroFormatted() {
+        // ---------------------------------------------------------
+        final ToDateString function = new ToDateString("yyyy-MM-dd HH:mm:ss.SSS");
+        // ---------------------------------------------------------
+
+        runExample(function, null, new Date(499165200L));
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ToDateStringExample.java]: Try different inputs for these. Try dates before 1970, nulls etc.
Same for the below examples

File: src/main/java/uk/gov/gchq/gaffer/doc/function/CsvToMapsExample.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Override
+    protected void runExamples() {
+        csvToMap();
+    }
+
+    private void csvToMap() {
+        // ---------------------------------------------------------
+        final CsvToMaps function = new CsvToMaps().header("header1", "header2", "header3").firstRow(1);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/CsvToMapsExample.java]: Try examples with empty csv values ie: ",,". Try examples with numbers and show what happens when it gets applied to nulls

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ParseDateExample.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        super(uk.gov.gchq.koryphe.impl.function.ParseDate.class);
+    }
+
+    @Override
+    protected void runExamples() {
+        parseDateGMT4();
+        parseDateGMT0();
+    }
+
+    private void parseDateGMT4() {
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ParseDateExample.java]: There may not be anything you can do but this title gets rendered to "Parse date g m t 4". It would be good if this read something like "Parse date four hours ahead of GMT"

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ParseDateExample.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        // ---------------------------------------------------------
+        final ParseDate parseDate = new ParseDate();
+        parseDate.setFormat("yyyy-MM-dd HH:mm:ss.SSS");
+        parseDate.setTimeZone("Etc/GMT+4");
+        // ---------------------------------------------------------
+
+        runExample(parseDate, "date string", "2015-10-21 16:29:00.000");
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ParseDateExample.java]: Could have more Date examples here

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ParseDateExample.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    private void parseDateGMT0() {
+        // ---------------------------------------------------------
+        final ParseDate parseDate = new ParseDate();
+        parseDate.setFormat("yyyy-MM-dd HH:mm");
+        parseDate.setTimeZone("Etc/GMT+0");
+        // ---------------------------------------------------------
+
+        runExample(parseDate, "date string", "2015-10-21 16:29");
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ParseDateExample.java]: Same as above

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ParseTimeExample.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        final ParseTime parseTime = new ParseTime();
+        // ---------------------------------------------------------
+        runExample(parseTime, null, "2000-01-02 03:04:05.006");
+    }
+
+    private void parseFormattedTime() {
+        // ---------------------------------------------------------
+        final ParseTime parseTime = new ParseTime().format("yyyy-MM hh:mm");
+        // ---------------------------------------------------------
+        runExample(parseTime, null, "2000-01-02 03:04");
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ParseTimeExample.java]: This throws an IllegalArgumentException. Although its good to include this, it would be good to include different working examples too as well as nulls.

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ParseTimeExample.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    private void parseFormattedGMT() {
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ParseTimeExample.java]: See previous comment on GMT in the subheadings

File: src/main/java/uk/gov/gchq/gaffer/doc/function/CsvLinesToMapsExample.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        final CsvLinesToMaps function = new CsvLinesToMaps()
+                .header("header1", "header2", "header3")
+                .firstRow(1);
+        // ---------------------------------------------------------
+
+        final List<String> input = Arrays.asList(
+                "header1,header2,header3",
+                "value1,value2,value3"
+        );
+        runExample(function, null, input);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/CsvLinesToMapsExample.java]: Add more inputs here

File: src/main/java/uk/gov/gchq/gaffer/doc/function/CsvLinesToMapsExample.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+                .header("header1", "header2", "header3")
+                .firstRow(1)
+                .delimiter('|');
+        // ---------------------------------------------------------
+
+        final List<String> input = Arrays.asList(
+                "header1|header2|header3",
+                "value1|value2|value3"
+        );
+        runExample(function, null, input);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/CsvLinesToMapsExample.java]: and here

File: src/main/java/uk/gov/gchq/gaffer/doc/function/DeserialiseJsonExample.java [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+        parseJson();
+    }
+
+    private void parseJson() {
+        // ---------------------------------------------------------
+        final DeserialiseJson function = new DeserialiseJson();
+        // ---------------------------------------------------------
+
+        final String input = "{\"elements\": [{\"value\": \"value1\"}, {\"value\": \"value2\"}]}";
+        runExample(function, null, input);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/DeserialiseJsonExample.java]: Add more inputs for this. Try complex objects (Entity Seeds, generics, Lists, etc)

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ApplyBiFunctionExample.java [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    protected void runExamples() {
+        applyBiFunctionSum();
+        applyBiFunctionMax();
+        applyBiFunctionMin();
+    }
+
+    private void applyBiFunctionSum() {
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ApplyBiFunctionExample.java]: This should be ApplyBiFunctionUsingSum or something like that.

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ApplyBiFunctionExample.java [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+        // ---------------------------------------------------------
+        final ApplyBiFunction<Number, Number, Number> function = new ApplyBiFunction<>(new Sum());
+        // ---------------------------------------------------------
+        final Tuple2<Number, Number> input = new Tuple2<>(1, 2);
+        runExample(function, null, input);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ApplyBiFunctionExample.java]: Just needs some more examples (same for below)

File: src/main/java/uk/gov/gchq/gaffer/doc/function/FunctionChainExample.java [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+    public FunctionChainExample() {
+        super(uk.gov.gchq.koryphe.impl.function.FunctionChain.class);
+    }
+
+    @Override
+    protected void runExamples() {
+        functionChain();
+    }
+
+    private void functionChain() {
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/FunctionChainExample.java]: This is quite a useful function so some more examples with different functions and inputs would be good

File: src/main/java/uk/gov/gchq/gaffer/doc/function/DeserialiseXmlExample.java [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    private void parseXML() {
+        // ---------------------------------------------------------
+        final DeserialiseXml function = new DeserialiseXml();
+        // ---------------------------------------------------------
+
+        final String input = "<root><element1><element2 attr=\"attr1\">value1</element2></element1><element1><element2>value2</element2></element1></root>";
+
+        runExample(function, null, input);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/DeserialiseXmlExample.java]: For this one, I'd just add a few more inputs. Maybe try some simpler ones as this can be quite difficult to read, for example:

Input Type | Input | Result Type | Result
-- | -- | -- | --
java.lang.String | &lt;root&gt;&lt;element1&gt;&lt;element2 attr="attr1"&gt;value1&lt;/element2&gt;&lt;/element1&gt;&lt;element1&gt;&lt;element2&gt;value2&lt;/element2&gt;&lt;/element1&gt;&lt;/root&gt; | java.util.HashMap | {root={element1=[{element2={attr=attr1, content=value1}}, {element2=value2}]}}

File: src/main/java/uk/gov/gchq/gaffer/doc/function/CurrentTimeExample.java (hunk 1 above)
Comment (GCHQDev404) [reply src/main/java/uk/gov/gchq/gaffer/doc/function/CurrentTimeExample.java]: These have empty input.

CancelScheduledJobExample.java
CountExample.java
CountGroupsExample.java
CurrentTimeExample.java
ExportToGafferResultCacheExample.java
ExportToSetExample.java
ForEachExample.java
GenerateElementsExample.java
GenerateObjectsExample.java
GetAdjacentIdsExample.java
GetAllElementsExample.java
GetAllJobDetailsExample.java
GetDataFrameOfElementsExample.java
GetElementsBetweenSetsExample.java
GetElementsExample.java
GetElementsInRangesExample.java
GetElementsWithinSetExample.java
GetGafferResultCacheExportExample.java
GetGraphFrameOfElementsExample.java
GetJavaRDDOfAllElementsExample.java
GetJavaRDDOfElementsExample.java
GetJobDetailsExample.java
GetJobResultsExample.java
GetSetExportExample.java
GetTraitsExample.java
JoinExample.java
LimitExample.java
MaxExample.java
MinExample.java
NamedOperationExample.java
ReduceExample.java
SortExample.java
ToArrayExample.java
ToCsvExample.java
ToEntitySeedsExample.java
ToListExample.java
ToMapExample.java
ToSetExample.java
ToSingletonListExample.java
ToStreamExample.java
ToVerticesExample.java

File: src/main/java/uk/gov/gchq/gaffer/doc/function/IncrementExample.java [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+        addToFloat();
+        addToLong();
+    }
+
+    private void addToInt() {
+        // ---------------------------------------------------------
+        final Increment increment = new Increment(3);
+        // ---------------------------------------------------------
+
+        runExample(increment, "return will match input type int", 2, 2.0d, 2.0f, 2L);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/IncrementExample.java]: For these strings, I think "The returned value type will match the input type Integer" sounds better.

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ToBytesExample.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+    protected void runExamples() {
+        stringToBytes();
+    }
+
+    private void stringToBytes() {
+        // ---------------------------------------------------------
+        final ToBytes toBytes = new ToBytes(StandardCharsets.UTF_16);
+        // ---------------------------------------------------------
+
+        runExample(toBytes, null, "example String", "", null, 1);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ToBytesExample.java]: Is there anything that can be done to make the output types and values more readable:

Input Type | Input | Result Type | Result
-- | -- | -- | --
java.lang.String | example String | [B | [B@31749b7a
java.lang.String |   | [B | [B@4b469e26
  | null |   | null
java.lang.Integer | 1 |   | ClassCastException: java.lang.Integer cannot be cast to java.lang.String

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ToTupleExample.java [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+        toTuple();
+    }
+
+    private void toTuple() {
+        // ---------------------------------------------------------
+        final ToTuple function = new ToTuple();
+        // ---------------------------------------------------------
+
+        Map<String, Object> input = new HashMap<>();
+        input.put("A", 1);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ToTupleExample.java]: This is fine except some of the inputs look a bit strange:

Input Type | Input | Result Type | Result
-- | -- | -- | --
java.util.ArrayList | [1, 2, 3, 4] | [java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer] | [1, 2, 3, 4]
[I | [I@6ae448f6 | [java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer] | [1, 2, 3, 4]
[Ljava.lang.Integer; | [1, 2, 3, 4] | [java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer] | [1, 2, 3, 4]
java.util.HashMap | {A=1, B=2, C=3} | uk.gov.gchq.koryphe.tuple.MapTuple | [1, 2, 3]

File: src/main/java/uk/gov/gchq/gaffer/doc/function/ParseDateExample.java [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+
+import uk.gov.gchq.koryphe.impl.function.ParseDate;
+
+public class ParseDateExample extends FunctionExample {
+    public static void main(final String[] args) {
+        new ParseDateExample().runAndPrint();
+    }
+
+    public ParseDateExample() {
+        super(uk.gov.gchq.koryphe.impl.function.ParseDate.class);
```
Comment (d47853) [inline src/main/java/uk/gov/gchq/gaffer/doc/function/ParseDateExample.java]:
```suggestion
        super(ParseDate.class);
```
